internal/vm/importers/glob.go

```diff
@@ -0,0 +1,233 @@
+func init() {
+	var ret []string
+	for k := range supportedParams {
+		ret = append(ret, k)
+	}
+	sort.Strings(ret)
+	supportedParamsDisplay = strings.Join(ret, ", ")
+}
+
+// globImportParams are additional parameters for the glob import expressed as URI query params.
+type globImportParams struct {
+	verb           string // one of 'import' or 'importstr'
+	dirs           int    // number of directories levels to retain in the returned object (default: all)
+	stripExtension bool   // whether extensions should be stripped
+}
+
+// keyFor returns the key to be used for the supplied file based on whether directory levels should be limited
+// and/ or extensions stripped.
+func (p globImportParams) keyFor(file string) (string, error) {
+	// short-circuit simple case
+	if p.dirs < 0 && !p.stripExtension {
+		return file, nil
+	}
+	elements := strings.Split(file, string(filepath.Separator))
+	name := elements[len(elements)-1]
+	dirs := elements[0 : len(elements)-1]
+	if p.dirs >= 0 {
+		if len(dirs) > p.dirs {
+			dirs = dirs[len(dirs)-p.dirs:]
+		}
+	}
+	if p.stripExtension {
+		pos := strings.LastIndex(name, ".")
+		if pos > 0 {
+			name = name[:pos]
+		}
+	}
+	finalElements := append(dirs, name)
+	return filepath.Join(finalElements...), nil
+}
+
+func newGlobParams(query url.Values) (params globImportParams, err error) {
+	for k := range query {
+		if !supportedParams[k] {
+			return params, fmt.Errorf("invalid query parameter '%s', allowed values are: %s", k, supportedParamsDisplay)
+		}
+	}
+	verb := query.Get(globParamVerb)
+	if verb == "" {
+		verb = "import"
+	}
+	if verb != "import" && verb != "importstr" {
+		return params, fmt.Errorf("'%s' parameter for glob import must be one of 'import' or 'importstr', found '%s'", globParamVerb, verb)
+	}
+	dirs := -1
+	dirsStr := query.Get(globParamDirs)
+	if dirsStr != "" {
+		l, err := strconv.Atoi(dirsStr)
+		if err != nil {
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, %v", dirsStr, globParamDirs, err)
+		}
+		if l < 0 {
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, must be non-negative", globParamDirs, dirsStr)
+		}
+		dirs = l
+	}
+	stripExtension := false
+	stripStr := query.Get(globParamStripExtension)
+	if stripStr != "" {
+		switch stripStr {
+		case "false":
+		case "true":
+			stripExtension = true
+		default:
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, must be 'true' or 'false'", stripStr, globParamStripExtension)
+		}
+	}
+	return globImportParams{
+		verb:           verb,
+		dirs:           dirs,
+		stripExtension: stripExtension,
+	}, nil
+}
+
+// globEntry is a cache entry for a glob path resolved from the current working directory.
+type globEntry struct {
+	contents jsonnet.Contents
+	foundAt  string
+	err      error
+}
+
+// GlobImporter provides facilities to import a bag of files using a glob pattern. Note that it will NOT
+// honor any library paths and must be exactly resolved from the caller's location.
```

Comment:
I am curious as to why we would not want to support library paths?

Reply:
because when you say `import 'foo.libsonnet'` the meaning can be easily defined to be "import foo.libsonnet in current directory and, if not found, the first file called foo.libsonnet in the path".

But, when you say `import 'glob:*.libsonnet' it could mean one of 3 things

* import all libsonnet files in current directory
* import all libsonnet files from the first dir in the path that contains_any_ libsonnet files
* import all libsonnet files from current directory and everywhere else in the PATH

So, the semantics are: import only the files relative to the current path and completely ignore jpaths (the first interpretation).

Reply:
Makes sense. Thanks

Reply:
Hmmm... there's some subtlety here. This won't honor JPath when resolving globs, but the paths to invidividual paths are resolved normally. This shouldn't be a problem, because JPath is checked after normal files, so it should behave as expected. I see only one way to trigger this issue and that is to have a file named "glob:somethingsomething". CompositeImporter will then try to use globimporter again to handle it. This sounds like a very weird boundary case, though (might be worth documenting).